AP_RSSI & GCS_Mavlink: Make rssi adhere to MAVLink standard #2

olliw42 opened the pull request:
continuation of https://github.com/ArduPilot/ardupilot/pull/18146

rebased and clean up as suggested

khancyr reviewed (approved):
LGTM
Thanks !

peterbarker commented on libraries/AP_RSSI/AP_RSSI.h:
```diff
@@ -53,7 +53,10 @@ class AP_RSSI
     // 0 represents weakest signal, 255 represents maximum signal.
-    uint8_t read_receiver_rssi_uint8();   
+    uint8_t read_receiver_rssi_uint8();
+    // Read the receiver RSSI value as an 8-bit integer with adhering to MAVLink standard
+    // 0 represents weakest signal, 254 represents maximum signal, 255 represents invalid/unknown signal
+    uint8_t read_receiver_rssi_mavlink();
```
This should be `GCS_MAVLINK::receiver_rssi()` - AP_RSSI shouldn't know anything about how mavlink works.

peterbarker reviewed (changes requested):
Scaling is still wrong - as you mentioned in the closed PR, 255 should be excluded.

Also, while it takes advantage of the range afforded by the type and description, the spec doesn't say you should do that but rather uselessly says units and scale are "sensor dependent".

MissionPlanner is scaling based on 0-255 (ignoring 255 meaning "no value").

MAVProxy isn't scaling the value, just presenting it (and ignoring the flag value).

QGC has special code which rescales ArduPilot's from 0-255 to 0-100.  It ignores the flag value.

PX4 Firmware seems to output 0-100.  It does not send the flag value for no rssi AFAICS, rather 0.

I do like sending through the flag value like this - but we'd be the only larger player in the ecosystem to do it AFAICS.

olliw42 commented:
and why should other libraries&consumers like LTM etx care about how mavlink works ... so two types of code are clearly needed, your comments appear a bit contradictory
I have not made the mavlink standard here and what you say seem to imply that it doesn't reflect reality, it's then a different playing field with other forces involved
I think I want to close this, it becomes nonsensical

peterbarker commented:
On Sun, 8 Aug 2021, olliw42 wrote:

> and why should other libraries&consumers like LTM etx care about how mavlink works ... so two types of code are clearly needed, your comments appear a bit contradictory

I'm not sure which side you're arguing here.  You're adding a method into 
AP_RSSI which is mavlink-specific - and then mentioning that LTM shouldn't 
really be seeing such a method?  I agree with that sentiment, which is why 
I'm suggesting the mavlink-specific method be defined on the mavlink 
library.

olliw42 commented:
I haven't really understood what you were saying
so, for dummies, I should move the mavlink specific function out of AP_RSSI but into GCS_MAVLINK, is this it?

concerning scaling, could you pl make a definite statement how you want it to be

concerning UINT16_MAX, could you pl make a definite statement how you want it to be

it's a bit confusing and makes me feel "what the heck am I supposed to do now" when interesting thoughts are presented without a definite conclusion however ;)

peterbarker commented:
On Sun, 8 Aug 2021, olliw42 wrote:

> I haven't really understood what you were saying
> so, for dummies, I should move the mavlink specific function out of AP_RSSI but into GCS_MAVLINK, is this it?

Yes, that's what I was trying to get at.

> concerning scaling, could you pl make a definite statement how you want it to be

I'll bring it up on a DevCall.  I'm just opinion, will be good to get 
consensus.

tridge commented:
we're going to get support headaches if we change the scaling when existing tools depend on the old scaling. It's fine to reserve 255, so scaling changes by 0.5%, but changing scaling by more than 50% will lead to worried users

khancyr commented:
I didn't follow on devcall , but where the 50% comes from ?

This is only changing the 255 to be excluded from the rssi scaling. Previously we were mapping the 100% rssi to 255

olliw42 commented:
concerning scaling, we talk about two scenarios, in my understanding:

scenario A:
```
rssi = read_receiver_rssi() * 255;
return (rssi < 255) ? rssi : 254;
```

scenario B:
```
return read_receiver_rssi() * 254;
```

In scenario A, the rssi values would be equal across the different telemetries (and logs?), except of 255 which is excluded, i.e. the range of 0 ... 254 would not exactly be 0 ... 1, but only 0 ... 0.9961

In scenario B the rssi would be 0.39% smaller than different telemetries (and logs?), but the range of 0 ... 254 would exactly correspond to 0 ... 1.

olliw42 commented:
>> On Sun, 8 Aug 2021, olliw42 wrote: I haven't really understood what you were saying so, for dummies, I should move the mavlink specific function out of AP_RSSI but into GCS_MAVLINK, is this it?

> Yes, that's what I was trying to get at.

got it, sorry for being so slow on this

>> concerning scaling, could you pl make a definite statement how you want it to be

> I'll bring it up on a DevCall. I'm just opinion, will be good to get consensus.

great, many thx
yes, would be good to have unambiguous consensus

peterbarker commented:
On Wed, 11 Aug 2021, olliw42 wrote:

> yes, would be good to have unambiguous consensus

For the first time in human history? :-)

So the upshot is:
  - scale 0-254 (a pure float value of 1 from an RSSI backend could easily 
happen)
  - 255 is our flag value.  We will need to change the GCSs to understand 
it.




p.s. the 50% scaling thing comes from a discussion to see if we could be 
consistent with other flight stacks in terms of scaling in this field. 
Sadly, the answer would appear to be "can't get to there without breaking 
everything")

olliw42 commented:
sorry for my silence, I was few days in vacation and then had to workout the github authentication thing LOL

@peterbarker 

> So the upshot is:
>  - scale 0-254 (a pure float value of 1 from an RSSI backend could easily happen)
>  - 255 is our flag value.  We will need to change the GCSs to understand it.

I'm sorry, I yet again seem to not understand: Is this now a decision for scenario A or scenario B (as defined in https://github.com/ArduPilot/ardupilot/pull/18244#issuecomment-896804440), or is this still undecided?

olliw42 commented:
I have done the move to the GCS_MAVLINK class as suggested. It in deed makes it looking better and was a good idea.

It was bit more difficult though as thought, since the send_servo_out() method was declared in the Rover class but not the GCS_MAVLINK_Rover class. So I moved it from Rover to GCS_MAVLINK_Rover. This does make lots of sense to me, and I have checked that this does compile for rover, but I could NOT check if that's indeed working!! So I kindly ask to review this change carefully.

The comm,it history may not look ok now. If so, I will do a 3rd and cleaned PR once the scaleing thing has been settled and the code is acceptable.

peterbarker commented:
Looking really good!  Our CI is sufficient testing for the servo-out change.

Just need the commitlist fixed, I think.

olliw42 commented:
so, this means the scaling should be as is?
could you pl drop a simple yes or no, I feel a bit unhappy with having to guess

olliw42 commented:
would it be possible to get a final verdict on the scaling?
I could then finalize this :)

olliw42 commented:
continued here https://github.com/ArduPilot/ardupilot/pull/18435
hence closed
